packages/go_router/lib/src/delegate.dart

```diff
@@ -146,22 +146,38 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   }
 
   bool _onPopPage(Route<Object?> route, Object? result) {
-    if (!route.didPop(result)) {
-      return false;
-    }
     final Page<Object?> page = route.settings as Page<Object?>;
     final RouteMatch? match = builder.getRouteMatchForPage(page);
     assert(match != null);
+    if (match!.route.onExit != null) {
+      // This function needs to return before calling onExit.
+      Future<void>.microtask(() async {
+        final bool result =
+            await match.route.onExit!(navigatorKey.currentContext!);
```

Comment:
if onExit returns FutureOr, and it does return synchronously, we can still call route.didpop in _onPopPage

Reply:
Changed it in [Make onExit return FutureOr](https://github.com/flutter/packages/pull/3709/commits/70d2440fd51c9be6153912eb629a1b757194cbb0)

Reply:
And if onExit is asynchronous, we don't call `route.didPop`?